feat(dart): Add a dev-mode check for undeclared lifecycle interfaces

kegluneq opened the pull request:
Add a check in `ReflectionCapabilities#interfaces` which determines if
the passed-in type implements a Lifecycle Interface but does not declare
that it does so.

See https://goo.gl/b07Kii for details.

Closes #6849

tbosch commented on modules/angular2/test/core/reflection/reflector_spec.ts:
```diff
@@ -214,6 +221,10 @@ export function main() {
           expect(p).toEqual([]);
         });
+
+        it("should throw for undeclared lifecycle interfaces", () => {
```
Maybe one more testcase with a class that just has an `ngOnOnit` method, but no `OnInit` interfaces in the class nor in the super class?

kegluneq commented:
Additional test added - thanks @tbosch !

mary-poppins commented:
Merging PR #6879 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-6879](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-6879).

mary-poppins commented:
Merging PR #6879 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-6879](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-6879).

mary-poppins commented:
Merging PR #6879 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-6879](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-6879).

vikerman commented:
Manually merged into master

vikerman commented:
Reverted from master - This needs more co-ordination on google3 side.

martine commented:
Got a pointer to what additional coordination is needed?

vikerman commented:
I couldn't get this last week - https://github.com/angular/angular/wiki/Angular-2-Dart-Transformer#lifecycle-hooks-are-not-being-called . We need to refactor google3 before landing this.

kara commented:
@kegluneq Looks like there are some failing Travis runs. Can you take a look?

kegluneq commented:
@kara looks like all Travis runs are failing with this error:

```bash
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!  Your environment is not in a good shape. Following issues were found:
!!!   - Your node_modules directory is stale or out of sync with npm-shrinkwrap.json. Run: npm install
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Error: Cannot find module 'broccoli-file-creator'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/home/travis/build/angular/angular/dist/tools/broccoli/trees/node_tree.js:10:17)
    at Module._compile (module.js:397:26)
    at Object.Module._extensions..js (module.js:404:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
```

Since this PR doesn't touch any Travis or Node settings, I don't think it's the culprit.

Commit:
feat(dart): Add a dev-mode check for undeclared lifecycle interfaces

Add a check in ReflectionCapabilities#interfaces which determines if
the passed-in type implements a Lifecycle Interface but does not declare
that it does so.

See https://goo.gl/b07Kii for details.

Closes #6849

kegluneq commented:
On second thought, could this be caused by how old this PR is? I'll rebase to master & re-push to see if things clear up.

kegluneq commented:
@kara, rebaseing to master & re-pushing seemed to work - this is now green again on Travis

kara commented:
Excellent, merging
